### PR TITLE
refactor: commit function

### PR DESCRIPTION
**Pull request by aner-starkware**
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/committer/297)

### CODE REVIEW

**Review by aner-starkware — COMMENTED** [review]
Python side: 
https://reviewable.io/reviews/starkware-industries/starkware/35461#-

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/committer/297#-:-O1RT62k9AEXqmkFyPzi:b9xki74)* status: 0 of 4 files reviewed, all discussions resolved (waiting on @TzahiTaub and @yoavGrs)

**Review by TzahiTaub — CHANGES_REQUESTED** [review]
> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/committer/297)* status: all files reviewed, 2 unresolved discussions (waiting on @aner-starkware and @yoavGrs)

___
*[`crates/committer_cli/src/commands.rs` line 10 at r1](https://reviewable.io/reviews/starkware-libs/committer/297#-O1S2rIg2iZEcioy-b0_:-O1S2rIg2iZEcioy-b0a:bwgdwil) ([raw file](https://github.com/starkware-libs/committer/blob/c49d7b2ffa24bbcab4839de4c8648c636c8b8cfc/crates/committer_cli/src/commands.rs#L10)):*
> ```XML
> };
> 
> pub async fn commit(input: Input<ConfigImpl>, output_path: String) {
> ```

I don't know if it makes more sense for this to be `Input` or `RawInput`so I'm just raising the question @dorimedini-starkware 



_Code quote:_
```XML
commit(input: Input<ConfigImpl>
```
___
*[`crates/committer_cli/src/tests/regression_tests.rs` line 50 at r1](https://reviewable.io/reviews/starkware-libs/committer/297#-O1S4AjS7XNcIvt-BF26:-O1S4AjS7XNcIvt-BF27:b-vea1ro) ([raw file](https://github.com/starkware-libs/committer/blob/c49d7b2ffa24bbcab4839de4c8648c636c8b8cfc/crates/committer_cli/src/tests/regression_tests.rs#L50)):*
> ```XML
>         ))
>     }
> }
> ```

IIUC the purpose of this wrapper, can't you add something like 
`#[cfg_attr(test, derive(Deserialize))]` above the `Input` declaration and remove this struct?



_Code quote:_
```XML
struct CommitterInput(Input<ConfigImpl>);

impl<'de> Deserialize<'de> for CommitterInput {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        Ok(Self(
            parse_input(&String::deserialize(deserializer)?).unwrap(),
        ))
    }
}
```

**Review by dorimedini-starkware — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/committer/297)* status: all files reviewed, 2 unresolved discussions (waiting on @aner-starkware, @TzahiTaub, and @yoavGrs)

___
*[`crates/committer_cli/benches/committer_bench.rs` line 69 at r1](https://reviewable.io/reviews/starkware-libs/committer/297#-O1SBjpb2oc464JTvXft:-O1SBjpb2oc464JTvXfu:bxlme3f) ([raw file](https://github.com/starkware-libs/committer/blob/c49d7b2ffa24bbcab4839de4c8648c636c8b8cfc/crates/committer_cli/benches/committer_bench.rs#L69)):*
> ```XML
>     criterion.bench_function("full_committer_flow", |benchmark| {
>         benchmark.iter(|| {
>             // TODO(Aner, 10/07/2024): replace functions with running main.
> ```

WDYM? you want to open a subprocess that runs the committer_cli?



_Code quote:_
```XML
// TODO(Aner, 10/07/2024): replace functions with running main.
```
___
*[`crates/committer_cli/src/commands.rs` line 10 at r1](https://reviewable.io/reviews/starkware-libs/committer/297#-O1S2rIg2iZEcioy-b0_:-O1SBsqb6VuB9wIixibl:b-nody) ([raw file](https://github.com/starkware-libs/committer/blob/c49d7b2ffa24bbcab4839de4c8648c636c8b8cfc/crates/committer_cli/src/commands.rs#L10)):*
<details><summary><i>Previously, TzahiTaub (Tzahi) wrote…</i></summary><blockquote>

I don't know if it makes more sense for this to be `Input` or `RawInput`so I'm just raising the question @dorimedini-starkware
</blockquote></details>

I think we should separate deserialization of input from the actual business logic, so I'm in favor of `Input`
___
*[`crates/committer_cli/src/tests/regression_tests.rs` line 50 at r1](https://reviewable.io/reviews/starkware-libs/committer/297#-O1S4AjS7XNcIvt-BF26:-O1SC7H02NLGoYPUgFxt:b-pdwza7) ([raw file](https://github.com/starkware-libs/committer/blob/c49d7b2ffa24bbcab4839de4c8648c636c8b8cfc/crates/committer_cli/src/tests/regression_tests.rs#L50)):*
<details><summary><i>Previously, TzahiTaub (Tzahi) wrote…</i></summary><blockquote>

IIUC the purpose of this wrapper, can't you add something like 
`#[cfg_attr(test, derive(Deserialize))]` above the `Input` declaration and remove this struct?
</blockquote></details>

I also don't think you need a wrapper; just `impl<'de, T: Config> Deserialize<'de> for Input<T> { ... }`

**Review by TzahiTaub — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/committer/297)* status: all files reviewed, 2 unresolved discussions (waiting on @dorimedini-starkware and @yoavGrs)

___
*[`crates/committer_cli/benches/committer_bench.rs` line 69 at r1](https://reviewable.io/reviews/starkware-libs/committer/297#-O1SBjpb2oc464JTvXft:-O1VnUnBDo_0biZx4iAf:b-e3nz8f) ([raw file](https://github.com/starkware-libs/committer/blob/c49d7b2ffa24bbcab4839de4c8648c636c8b8cfc/crates/committer_cli/benches/committer_bench.rs#L69)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

WDYM? you want to open a subprocess that runs the committer_cli?
</blockquote></details>

I think we discussed about calling the `main` function of the committer_cli directly.

**Review by dorimedini-starkware — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/committer/297)* status: all files reviewed, 2 unresolved discussions (waiting on @TzahiTaub and @yoavGrs)

___
*[`crates/committer_cli/benches/committer_bench.rs` line 69 at r1](https://reviewable.io/reviews/starkware-libs/committer/297#-O1SBjpb2oc464JTvXft:-O1Voy5YBub-6Olv_wyY:b-ap16l5) ([raw file](https://github.com/starkware-libs/committer/blob/c49d7b2ffa24bbcab4839de4c8648c636c8b8cfc/crates/committer_cli/benches/committer_bench.rs#L69)):*
<details><summary><i>Previously, TzahiTaub (Tzahi) wrote…</i></summary><blockquote>

I think we discussed about calling the `main` function of the committer_cli directly.
</blockquote></details>

I don't think that's good practice; it's the entry point of the binary (only argument parsing and logger initialization). the contents of this `main` is only this:
```
            let input_string = read_from_stdin();
            commit(&input_string, output_path).await;
```
... so if you want to include input deserialization, just extract these two lines into a function, and call it

**Review by dorimedini-starkware — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/committer/297)* status: all files reviewed, 2 unresolved discussions (waiting on @TzahiTaub and @yoavGrs)

___
*[`crates/committer_cli/benches/committer_bench.rs` line 69 at r1](https://reviewable.io/reviews/starkware-libs/committer/297#-O1SBjpb2oc464JTvXft:-O1VpYymDP6NEh5DVg3c:bhs4b3m) ([raw file](https://github.com/starkware-libs/committer/blob/c49d7b2ffa24bbcab4839de4c8648c636c8b8cfc/crates/committer_cli/benches/committer_bench.rs#L69)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

I don't think that's good practice; it's the entry point of the binary (only argument parsing and logger initialization). the contents of this `main` is only this:
```
            let input_string = read_from_stdin();
            commit(&input_string, output_path).await;
```
... so if you want to include input deserialization, just extract these two lines into a function, and call it
</blockquote></details>

... but since you're not using stdin, I see no reason to do even that

**Review by dorimedini-starkware — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/committer/297)* status: all files reviewed, 2 unresolved discussions (waiting on @aner-starkware, @TzahiTaub, and @yoavGrs)

___
*[`crates/committer_cli/benches/committer_bench.rs` line 69 at r1](https://reviewable.io/reviews/starkware-libs/committer/297#-O1SBjpb2oc464JTvXft:-O1VsuIT8b6p4YgbQ_fi:bdmrv26) ([raw file](https://github.com/starkware-libs/committer/blob/c49d7b2ffa24bbcab4839de4c8648c636c8b8cfc/crates/committer_cli/benches/committer_bench.rs#L69)):*
<details><summary><i>Previously, aner-starkware wrote…</i></summary><blockquote>

I think the logic is that:
1. If someone changes main (e.g., using a different commit according some flag), these changes are automatically transferred to this benchmarking, they need to be written manually.
2. We're not benchmarking exactly the same thing. For example, we don't time reading from std_in (we assume that read_from_stdin() is cheap, but we might be missing something like when we wrote to file wrong).
</blockquote></details>

if you want to invoke the CLI *binary* (i.e. subprocess?), then using main is good.
because of (1) and (2), this is probably a good idea :)
but no reason to call the `main` function from within rust IMO, unless you want to pass arguments (which is the point of the main function).

**Comment by aner-starkware** [comment]
*[`crates/committer_cli/src/tests/regression_tests.rs` line 50 at r1](https://reviewable.io/reviews/starkware-libs/committer/297#-O1S4AjS7XNcIvt-BF26:-O1VtRziERb2j6v6j4sn:beg49jl) ([raw file](https://github.com/starkware-libs/committer/blob/c49d7b2ffa24bbcab4839de4c8648c636c8b8cfc/crates/committer_cli/src/tests/regression_tests.rs#L50)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

I also don't think you need a wrapper; just `impl<'de, T: Config> Deserialize<'de> for Input<T> { ... }`
</blockquote></details>

This cannot be done without a larger refactoring, because Input<T> is in a different crate (`committer`) than the function that deserializes it (`parse_input` in `committer_cli`).

**Comment by aner-starkware** [comment]
*[`crates/committer_cli/benches/committer_bench.rs` line 69 at r1](https://reviewable.io/reviews/starkware-libs/committer/297#-O1SBjpb2oc464JTvXft:-O1VuoTlEp97lPVm_Psm:bjtyqsq) ([raw file](https://github.com/starkware-libs/committer/blob/c49d7b2ffa24bbcab4839de4c8648c636c8b8cfc/crates/committer_cli/benches/committer_bench.rs#L69)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

if you want to invoke the CLI *binary* (i.e. subprocess?), then using main is good.
because of (1) and (2), this is probably a good idea :)
but no reason to call the `main` function from within rust IMO, unless you want to pass arguments (which is the point of the main function).
</blockquote></details>

I don't understand

**Review by dorimedini-starkware — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/committer/297)* status: all files reviewed, 2 unresolved discussions (waiting on @TzahiTaub and @yoavGrs)

___
*[`crates/committer_cli/src/tests/regression_tests.rs` line 50 at r1](https://reviewable.io/reviews/starkware-libs/committer/297#-O1S4AjS7XNcIvt-BF26:-O1VuNSR2l7XjbeqOXnH:beo91sx) ([raw file](https://github.com/starkware-libs/committer/blob/c49d7b2ffa24bbcab4839de4c8648c636c8b8cfc/crates/committer_cli/src/tests/regression_tests.rs#L50)):*
<details><summary><i>Previously, aner-starkware wrote…</i></summary><blockquote>

This cannot be done without a larger refactoring, because Input<T> is in a different crate (`committer`) than the function that deserializes it (`parse_input` in `committer_cli`).
</blockquote></details>

ah... right, cannot `impl T for S` if both `T` and `S` are defined outside the current crate.
your solution (wrapping `Input` in a type defined within the crate) seems like the good one

**Review by dorimedini-starkware — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/committer/297)* status: all files reviewed, 2 unresolved discussions (waiting on @aner-starkware, @TzahiTaub, and @yoavGrs)

___
*[`crates/committer_cli/benches/committer_bench.rs` line 69 at r1](https://reviewable.io/reviews/starkware-libs/committer/297#-O1SBjpb2oc464JTvXft:-O1Vuwp28fosz-MZ1MmC:b-e2vxco) ([raw file](https://github.com/starkware-libs/committer/blob/c49d7b2ffa24bbcab4839de4c8648c636c8b8cfc/crates/committer_cli/benches/committer_bench.rs#L69)):*
<details><summary><i>Previously, aner-starkware wrote…</i></summary><blockquote>

I don't understand
</blockquote></details>

I agree with you that we should run `main` here.
however, you shouldn't "call `main`" directly from rust code, you should be building the release binary and *running* it.
if this is impossible, then I don't think we can do better than the current solution: `main` may change but it's logic is mainly for argument parsing anyway, which is only relevant if you run the binary...
tl;dr: I would keep the code as-is for now, and in subsequent work see if the binary itself can be run from the context of the benchmark

**Comment by aner-starkware** [comment]
*[`crates/committer_cli/src/tests/regression_tests.rs` line 50 at r1](https://reviewable.io/reviews/starkware-libs/committer/297#-O1S4AjS7XNcIvt-BF26:-O1VwYdo1YppyUkbqTJY:b6lpskq) ([raw file](https://github.com/starkware-libs/committer/blob/c49d7b2ffa24bbcab4839de4c8648c636c8b8cfc/crates/committer_cli/src/tests/regression_tests.rs#L50)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

ah... right, cannot `impl T for S` if both `T` and `S` are defined outside the current crate.
your solution (wrapping `Input` in a type defined within the crate) seems like the good one
</blockquote></details>

It would be good if it weren't for the annoying `.0` that needs to be added...

**Review by dorimedini-starkware — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/committer/297)* status: all files reviewed, 1 unresolved discussion (waiting on @TzahiTaub and @yoavGrs)

___
*[`crates/committer_cli/src/tests/regression_tests.rs` line 50 at r1](https://reviewable.io/reviews/starkware-libs/committer/297#-O1S4AjS7XNcIvt-BF26:-O1VxlILETR3AojHtf_G:b-ytd5x2) ([raw file](https://github.com/starkware-libs/committer/blob/c49d7b2ffa24bbcab4839de4c8648c636c8b8cfc/crates/committer_cli/src/tests/regression_tests.rs#L50)):*
<details><summary><i>Previously, aner-starkware wrote…</i></summary><blockquote>

It would be good if it weren't for the annoying `.0` that needs to be added...
</blockquote></details>

if you add `derive_more::Deref` this issue should go away usually

**Comment by aner-starkware** [comment]
*[`crates/committer_cli/src/tests/regression_tests.rs` line 50 at r1](https://reviewable.io/reviews/starkware-libs/committer/297#-O1S4AjS7XNcIvt-BF26:-O1W2MhF4STG8nRjlejE:b-5h3p3k) ([raw file](https://github.com/starkware-libs/committer/blob/c49d7b2ffa24bbcab4839de4c8648c636c8b8cfc/crates/committer_cli/src/tests/regression_tests.rs#L50)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

if you add `derive_more::Deref` this issue should go away usually
</blockquote></details>

How do you add it? I couldn't figure it out. But I implemented `Deref` directly, is it OK?

**Review by dorimedini-starkware — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/committer/297)* status: 2 of 4 files reviewed, 2 unresolved discussions (waiting on @aner-starkware, @TzahiTaub, and @yoavGrs)

___
*[`crates/committer_cli/src/tests/regression_tests.rs` line 50 at r1](https://reviewable.io/reviews/starkware-libs/committer/297#-O1S4AjS7XNcIvt-BF26:-O1WCBeN2yMVB09OxmEe:bsd42l6) ([raw file](https://github.com/starkware-libs/committer/blob/c49d7b2ffa24bbcab4839de4c8648c636c8b8cfc/crates/committer_cli/src/tests/regression_tests.rs#L50)):*
<details><summary><i>Previously, aner-starkware wrote…</i></summary><blockquote>

How do you add it? I couldn't figure it out. But I implemented `Deref` directly, is it OK?
</blockquote></details>

yes, but what I suggested doesn't work...? non-blocking
___
*[`crates/committer_cli/src/tests/regression_tests.rs` line 34 at r2](https://reviewable.io/reviews/starkware-libs/committer/297#-O1WC6GFEOuHH9YFlp4D:-O1WC6GFEOuHH9YFlp4E:b-mn45u5) ([raw file](https://github.com/starkware-libs/committer/blob/e008a07f637ab0ce1a7a8217322fbb2e202f412c/crates/committer_cli/src/tests/regression_tests.rs#L34)):*
> ```XML
>         &self.0
>     }
> }
> ```

this doesn't work?



_Suggestion:_
```XML
#[derive(derive_more::Deref)]
struct FactMap(Map<String, Value>);
```

**Review by aner-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/committer/297)* status: 2 of 4 files reviewed, 2 unresolved discussions (waiting on @dorimedini-starkware, @TzahiTaub, and @yoavGrs)

___
*[`crates/committer_cli/src/tests/regression_tests.rs` line 50 at r1](https://reviewable.io/reviews/starkware-libs/committer/297#-O1S4AjS7XNcIvt-BF26:-O1WDXPx8xmS5FUULNvM:b-896fix) ([raw file](https://github.com/starkware-libs/committer/blob/c49d7b2ffa24bbcab4839de4c8648c636c8b8cfc/crates/committer_cli/src/tests/regression_tests.rs#L50)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

yes, but what I suggested doesn't work...? non-blocking
</blockquote></details>

Done.
___
*[`crates/committer_cli/src/tests/regression_tests.rs` line 34 at r2](https://reviewable.io/reviews/starkware-libs/committer/297#-O1WC6GFEOuHH9YFlp4D:-O1WDTAA7p71a9prGZHF:bc3c47p) ([raw file](https://github.com/starkware-libs/committer/blob/e008a07f637ab0ce1a7a8217322fbb2e202f412c/crates/committer_cli/src/tests/regression_tests.rs#L34)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

this doesn't work?
</blockquote></details>

Works, thanks :)

**Review by TzahiTaub — COMMENTED** [review]
> Reviewed 1 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/committer/297)* status: 3 of 4 files reviewed, all discussions resolved (waiting on @yoavGrs)

___
*[`crates/committer_cli/benches/committer_bench.rs` line 69 at r1](https://reviewable.io/reviews/starkware-libs/committer/297#-O1SBjpb2oc464JTvXft:-O1WwA2L8wqUHltModLi:b-8041cl) ([raw file](https://github.com/starkware-libs/committer/blob/c49d7b2ffa24bbcab4839de4c8648c636c8b8cfc/crates/committer_cli/benches/committer_bench.rs#L69)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

I agree with you that we should run `main` here.
however, you shouldn't "call `main`" directly from rust code, you should be building the release binary and *running* it.
if this is impossible, then I don't think we can do better than the current solution: `main` may change but it's logic is mainly for argument parsing anyway, which is only relevant if you run the binary...
tl;dr: I would keep the code as-is for now, and in subsequent work see if the binary itself can be run from the context of the benchmark
</blockquote></details>

Let's try to replace it as in the TODO in another PR.

**Review by TzahiTaub — APPROVED** [review]
